crates/radix-engine-toolkit/src/manifest_analysis/classification/simple_transfer.rs

```diff
@@ -0,0 +1,220 @@
+        self.transition(context.instruction());
+    }
+}
+
+impl ManifestAnalyzerRequirementState for SimpleTransferStateMachine {
+    fn requirement_state(&self) -> RequirementState {
+        match self {
+            SimpleTransferStateMachine::DepositPerformed => {
+                RequirementState::Fulfilled
+            }
+            SimpleTransferStateMachine::InitialState
+            | SimpleTransferStateMachine::AccessControllerProofCreated
+            | SimpleTransferStateMachine::FeeLockPerformed
+            | SimpleTransferStateMachine::ResourcesWithdrawn
+            | SimpleTransferStateMachine::ResourcesInBucket => {
+                RequirementState::CurrentlyUnfulfilled
+            }
+            SimpleTransferStateMachine::InvalidState => {
+                RequirementState::PermanentlyUnfulfilled
+            }
+        }
+    }
+
+    fn process_instruction(&mut self, context: AnalysisContext<'_>) {
+        self.transition(context.instruction());
+    }
+}
+
+impl ManifestAnalyzerPermissionState for SimpleTransferStateMachine {
+    fn all_instructions_permitted(&self) -> bool {
+        !matches!(self, Self::InvalidState)
+    }
+
+    fn process_instruction(&mut self, context: AnalysisContext<'_>) {
+        self.transition(context.instruction());
+    }
+}
+
+impl SimpleTransferStateMachine {
+    pub fn transition(&mut self, instruction: &GroupedInstruction) {
+        let next_state = match (&self, instruction) {
+            // Initial State -> Access Controller Proof State transition.
+            (
+                Self::InitialState,
+                GroupedInstruction::InvocationInstructions(
+                    InvocationInstructions::CallMethod(CallMethod {
+                        address: ManifestGlobalAddress::Static(address),
+                        method_name,
+                        ..
+                    }),
+                ),
+            ) if matches!(
+                address.as_node_id().entity_type().map(Into::into),
+                Some(GroupedEntityType::AccessControllerEntities(..))
+            ) && method_name == ACCESS_CONTROLLER_CREATE_PROOF_IDENT =>
+            {
+                Self::AccessControllerProofCreated
+            }
+            // (Initial State | Access Controller Proof Created) -> Lock Fee
+            // State transition.
+            (
+                Self::InitialState | Self::AccessControllerProofCreated,
+                GroupedInstruction::InvocationInstructions(
+                    InvocationInstructions::CallMethod(CallMethod {
+                        address: ManifestGlobalAddress::Static(address),
+                        method_name,
+                        ..
+                    }),
+                ),
+            ) if matches!(
+                address.as_node_id().entity_type().map(Into::into),
+                Some(GroupedEntityType::AccountEntities(..))
+            ) && method_name == ACCOUNT_LOCK_FEE_IDENT =>
```

Comment:
MINOR: This is repeated a lot, it'd be less noisy to review/understand if this was something like:

```rust
impl GroupedInstruction {
    fn static_only_invocation_matches(
        &self,
        by_entity: impl FnOnce(GroupedEntityType) -> bool,
        by_method: impl FnOnce(&str) -> bool,
    ) -> bool {
        // ...
    }

    fn invocation_matches(
        &self,
        context: AnalysisContext<'_>, // To map dynamic addresses
        by_entity: impl FnOnce(GroupedEntityType) -> bool,
        by_method: impl FnOnce(&str) -> bool,
    ) -> bool {
        // ...
    }
}

impl AnalysisContext {
    fn resolve_entity_type(&self, address: ManifestAddress) -> Option<GroupedEntityType>;
}
```

Reply:
Similar to the comment above, I've made a change where we move to use typed invocations which cleans this code up by a lot.